src/index/visitor.rs

```diff
@@ -653,3 +536,296 @@ fn visit_data_type(
         _ => { /* unnamed datatypes are ignored */ }
     };
 }
+
+/// Internally we create a fat pointer struct for VLAs, which consists of a pointer to the passed array plus
+/// its dimensions, such that `ARRAY[*, *, *] OF INT` becomes
+/// ```ignore
+/// STRUCT
+///     ptr       : REF_TO ARRAY[*] OF INT;
+///     dimensions: ARRAY[0..2, 0..1] OF DINT;
+///                       ^^^^       --> dimension index
+///                             ^^^^ --> start- & end-offset
+/// END_STRUCT
+/// ```
+fn visit_variable_length_array(
+    bounds: &AstStatement,
+    referenced_type: &DataTypeDeclaration,
+    name: &str,
+    index: &mut Index,
+    type_declaration: &UserTypeDeclaration,
+    symbol_location_factory: &SymbolLocationFactory,
+) {
+    let ndims = match bounds {
+        AstStatement::VlaRangeStatement { .. } => 1,
+        AstStatement::ExpressionList { expressions, .. } => expressions.len(),
+        _ => unreachable!("not a bounds statement"),
+    };
+
+    let referenced_type = referenced_type.get_name().expect("named datatype").to_string();
+    let struct_name = name.to_owned();
+
+    // register dummy array type so it can later be annotated as a type hint
+    let array_name = format!("__arr_vla_{ndims}_{referenced_type}").to_lowercase();
+    index.register_type(typesystem::DataType {
+        name: array_name.clone(),
+        initial_value: None,
+        information: DataTypeInformation::Array {
+            name: array_name.clone(),
+            inner_type_name: referenced_type.clone(),
+            // dummy dimensions that will never actually be used
+            dimensions: (0..ndims)
+                .map(|_| Dimension {
+                    start_offset: TypeSize::Undetermined,
+                    end_offset: TypeSize::Undetermined,
+                })
+                .collect::<Vec<_>>(),
+        },
+        nature: TypeNature::Any,
+        location: SymbolLocation::internal(),
+    });
+
+    // Create variable index entries for VLA struct members
+    let variables = vec![
+        // Pointer
+        Variable {
+            name: format!("struct_vla_{referenced_type}_{ndims}").to_lowercase(),
```

Comment:
should this typename also start with "__" so the user cannot create a conflicting type?